signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    @abstractmethod
+    def _load(self):
+        pass
+
+    @abstractmethod
+    def _sync(self):
+        pass
+
+    @contextmanager
+    def _safe_sync(self):
+        backup = self._data
+        try:
+            yield
+        except BaseException:
+            self._data = backup
+            raise
+
+    def sync(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._sync()
+            else:
+                self._parent.sync()
+
+    def load(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                data = self._load()
+                with self._suspend_sync():
+                    self.reset(data)
+            else:
+                self._parent.load()
+
+
+class _SyncedDict(SyncedCollection, MutableMapping):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
+
+    VALID_KEY_TYPES = (str, int, bool, type(None))
+
+    def __init__(self, **kwargs):
+        data = kwargs.pop('data', None)
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = {}
+        else:
+            self._data = {
+                self._validate_key(key): self.from_base(data=value, parent=self)
+                for key, value in data.items()
+            }
+
+    def to_base(self):
+        converted = {}
+        for key, value in self._data.items():
+            if isinstance(value, SyncedCollection):
+                converted[key] = value.to_base()
+            else:
+                converted[key] = value
+        return converted
+
+    def reset(self, data=None):
+        if data is None:
+            data = {}
+        backup = self._data
+        if isinstance(data, Mapping):
+            try:
+                with self._suspend_sync():
+                    for key in data:
+                        if key in self._data:
+                            if data[key] == self._data[key]:
+                                continue
+                            try:
+                                self._data[key].reset(key)
+                                continue
+                            except (ValueError, AttributeError):
+                                pass
+                        self._data[key] = self.from_base(data=data[key], parent=self)
+                    remove = set()
+                    for key in self._data:
+                        if key not in data:
+                            remove.add(key)
+                    for key in remove:
+                        del self._data[key]
+                self.sync()
+            except BaseException:  # rollback
+                self._data = backup
+                raise
+        else:
+            raise ValueError("The data must be a mapping or None not {}.".format(type(data)))
```

Comment:
This error might be more informative.
```suggestion
            raise ValueError("Unsupported type: {}. The data must be a mapping or None.".format(type(data)))
```